File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    private int res = 0;

    private int dfs(int i, int j, int[][] grid){
        if(i<0 || i>=grid.length || 
            j<0 || j>=grid[0].length){
                return 0;
            }

        if(grid[i][j] == 0){
            return 0;
        }
        

        grid[i][j] = 0;

        return 1 + dfs(i+1, j, grid) + dfs(i-1, j, grid) +
            dfs(i, j+1, grid) + dfs(i, j-1, grid);

    }

    public int maxAreaOfIsland(int[][] grid) {
        for(int i=0; i<grid.length; i++){
            for(int j=0; j<grid[0].length; j++){
                if(grid[i][j]==1){
                    res = Math.max(res, dfs(i,j,grid));
                }
            }
        }

        return res;
    }
}

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])

        def dfs(i, j):
            if i<0 or i>=m or j<0 or j>=n or grid[i][j]==0:
                return 0
            
            grid[i][j] = 0

            return 1 + dfs(i+1, j) + dfs(i-1, j) + dfs(i, j-1) + dfs(i, j+1)

        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    res = max(res, dfs(i,j))

        return res